internal/lock.go

```diff
@@ -49,30 +55,111 @@ func NewLock(path string, newOk bool) (*Lock, error) {
 	return &Lock{path: path, conf: conf}, nil
 }
 
-func (l *Lock) AddResource(paths []string, algo string, tags []string, filename string) error {
+func (l *Lock) AddResource(paths []string, algo string, tags []string, filename string, cacheURL string) error {
 	for _, u := range paths {
 		if l.Contains(u) {
 			return fmt.Errorf("resource '%s' is already present", u)
 		}
 	}
-	r, err := NewResourceFromUrl(paths, algo, tags, filename)
+	r, err := NewResourceFromUrl(paths, algo, tags, filename, cacheURL)
 	if err != nil {
 		return err
 	}
+
+	// If cache URL is provided, handles Artifactory upload
+	if cacheURL != "" {
+		token := os.Getenv("GRABIT_ARTIFACTORY_TOKEN")
+		if token == "" {
+			return fmt.Errorf("GRABIT_ARTIFACTORY_TOKEN environment variable is not set")
+		}
+
+		// Add context here
+		ctx := context.Background()
+		path, err := GetUrltoTempFile(paths[0], "", ctx) // Add empty string for bearer token
+		if err != nil {
+			return fmt.Errorf("failed to get file for cache: %s", err)
+		}
+		defer os.Remove(path)
+
+		// Upload to Artifactory using hash as filename
+		err = uploadToArtifactory(path, cacheURL, r.Integrity)
+		if err != nil {
+			return fmt.Errorf("failed to upload to cache: %v", err)
+		}
+	}
+
 	l.conf.Resource = append(l.conf.Resource, *r)
 	return nil
 }
 
+func uploadToArtifactory(filePath, cacheURL, integrity string) error {
+	token := os.Getenv("GRABIT_ARTIFACTORY_TOKEN")
+	if token == "" {
+		return fmt.Errorf("GRABIT_ARTIFACTORY_TOKEN environment variable is not set")
+	}
+
+	// Read the content of the file
+	fileContent, err := os.ReadFile(filePath)
+	if err != nil {
+		return fmt.Errorf("failed to read file: %v", err)
+	}
+
+	// Use the integrity value directly for the URL since it's already in the correct format
+	artifactoryURL := fmt.Sprintf("%s/%s", cacheURL, integrity)
```

Comment:
This is repeated in here and in DeleteResource and this is bad since these 2 calls absolutely need to be in sync. You need to refactor this into a utility method called in both places.

Reply:
working on address that